# **Hummingbird-ML**

[How to Harness GPU to Speed Up Machine Learning with Hummingbird-ML](https://www.youtube.com/watch?v=qN8jcUmo8TI)

Adapted from: https://github.com/microsoft/hummingbird

# Install Hummingbird-ML

In [ ]:
! pip install hummingbird-ml[extra]

     |████████████████████████████████| 61kB 1.8MB/s 
     |████████████████████████████████| 71kB 3.7MB/s 
     |████████████████████████████████| 7.4MB 16.0MB/s 


# Import libraries

In [ ]:
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from hummingbird.ml import convert

# Create some random data for binary classification

In [ ]:
num_classes = 2
X = np.random.rand(100000, 28)
y = np.random.randint(num_classes, size=100000)

# Create and train a model (scikit-learn RandomForestClassifier)

In [ ]:
skl_model = RandomForestClassifier(n_estimators=10, max_depth=10)

In [ ]:
%%timeit
skl_model.fit(X, y)

1 loop, best of 3: 4.78 s per loop


In [ ]:
%%timeit
skl_model.predict(X)

10 loops, best of 3: 85.6 ms per loop


# Use Hummingbird to convert the model to PyTorch

In [ ]:
model = convert(skl_model, 'pytorch')

# Run predictions on CPU

In [ ]:
%%timeit
model.predict(X)

1 loop, best of 3: 174 ms per loop


# Run predictions on GPU

In [ ]:
model.to('cuda')

In [ ]:
%%timeit
model.predict(X)

The slowest run took 5.22 times longer than the fastest. This could mean that an intermediate result is being cached.
100 loops, best of 3: 14.8 ms per loop


# Calculation Time

Methods | Timing | Performance
--|--|--
scikit-learn | 85.6 ms | -
PyTorch (CPU) | 174 ms | 2 X slower than scikit-learn
PyTorch (GPU) | 14.8 ms | Almost 6 X faster than scikit-learn; Almost 12 X faster than PyTorch (CPU)